automod/event.go

```diff
@@ -0,0 +1,273 @@
+package automod
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	comatproto "github.com/bluesky-social/indigo/api/atproto"
+	appbsky "github.com/bluesky-social/indigo/api/bsky"
+	"github.com/bluesky-social/indigo/atproto/identity"
+)
+
+type ModReport struct {
+	ReasonType string
+	Comment    string
+}
+
+// information about a repo/account/identity, always pre-populated and relevant to many rules
+type AccountMeta struct {
+	Identity *identity.Identity
+	// TODO: createdAt / age
+}
+
+// base type for events. events are both containers for data about the event itself (similar to an HTTP request type); aggregate results and state (counters, mod actions) to be persisted after all rules are run; and act as an API for additional network reads and operations.
+type Event struct {
+	Engine            *Engine
```

Comment:
it feels a bit odd to me that the event would carry the engine reference, seems like the 'persist' methods should just be on the engine itself and we should pass the event through that.

Reply:
yeah, agree this feels a bit weird.

the rules reach back to the engine other ways as well though: helpers on the event can fetch counters and things. it might be more golang-idiomatic to pass both the engine and the event to all rules.

part of this might be poor naming. "events" here are kind of more like `echo.Context`, maybe "EventContext" would be better.

might still be a bad design pattern.

Reply:
the more obvious/idiomatic thing would be to have the rules be methods on the engine, or methods on some context object in general. then you would access things like counters, logger, or other external resources through the parent struct. but i'm trying hard to make rules free-floating rules so they can be implemented in arbitrary packages and collected at configuration time.